src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleControl.cs

```diff
@@ -2575,19 +2568,29 @@ internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char
             }
 
             outBuffer = output.Slice(cursor);
+            if (!newLine)
+            {
+                WriteConsole(consoleHandle, outBuffer);
+                return;
+            }
 
-            if (newLine)
+            string lineEnding = Environment.NewLine;
+            int size = outBuffer.Length + lineEnding.Length;
+            if (size <= MaxStackAllocSize)
             {
-                var endOfLine = Environment.NewLine.AsSpan();
-                var endOfLineLength = endOfLine.Length;
-                Span<char> outBufferLine = stackalloc char[outBuffer.Length + endOfLineLength];
-                outBuffer.CopyTo(outBufferLine);
-                endOfLine.CopyTo(outBufferLine.Slice(outBufferLine.Length - endOfLineLength));
-                WriteConsole(consoleHandle, outBufferLine);
+                // We expect this code path to be hit more often, and thus optimize it with 'stackalloc'.
+                Span<char> buffer = stackalloc char[size];
+                outBuffer.CopyTo(buffer);
+                lineEnding.CopyTo(buffer.Slice(outBuffer.Length));
+                WriteConsole(consoleHandle, buffer);
             }
             else
             {
-                WriteConsole(consoleHandle, outBuffer);
+                // The size is 16K at most, so this array is a small object that will soon be collected.
+                var buffer = new char[size];
```

Comment:
Since we want to reduce GC pressure too we could use that .Net Runtime extensively utilize - ArrayPool.
Example https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/IO/File.cs,1166

In light of the previous comment about const stackalloc:
```c#
        internal static void WriteConsole(ConsoleHandle consoleHandle, ReadOnlySpan<char> output, bool newLine)
        {
            Dbg.Assert(!consoleHandle.IsInvalid, "ConsoleHandle is not valid");
            Dbg.Assert(!consoleHandle.IsClosed, "ConsoleHandle is closed");

            Span<char> buffer = stackalloc char[MaxStackAllocSize];

            ...

            byte[]? pooledBuffer = null;
            string lineEnding = Environment.NewLine;
            int size = outBuffer.Length + lineEnding.Length;
            buffer = size <= MaxStackAllocSize ? buffer.Slice(size) : pooledBuffer = ArrayPool<char>.Shared.Rent(size);

            try
            {
                outBuffer.CopyTo(buffer);
                lineEnding.CopyTo(sourceIndex: 0, buffer, outBuffer.Length, lineEnding.Length);
                WriteConsole(consoleHandle, buffer);
            }
            finally
            {
                if (pooledBuffer != null)
                {
                    ArrayPool<char>.Shared.Return(pooledBuffer);
                }
            }
```

Reply:
Thanks for bringing up `ArrayPool`. Code updated.